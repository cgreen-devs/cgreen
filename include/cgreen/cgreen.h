#include <stdlib.h>
#include <cgreen/unit.h>
#include <cgreen/suite.h>
#include <cgreen/text_reporter.h>
#include <cgreen/cdash_reporter.h>
#include <cgreen/cute_reporter.h>
#include <cgreen/assertions.h>
#include <cgreen/constraint_syntax_helpers.h>
#include <cgreen/runner.h>
#include <cgreen/boxed_double.h>
