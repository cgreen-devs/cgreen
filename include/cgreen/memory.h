#ifndef MEMORY_HEADER
#define MEMORY_HEADER

#ifndef __cplusplus
void* cgreen_malloc(size_t size, const char * filename, int line);
#define malloc(size) cgreen_malloc(size, __FILE__, __LINE__)

void* cgreen_calloc(size_t num_members, size_t size, const char * filename, int line);
#define calloc(num_members, size) cgreen_calloc(num_members, size, __FILE__, __LINE__)

void* cgreen_realloc(void* pointer, size_t size, const char * filename, int line);
#define realloc(pointer, size) cgreen_realloc(pointer, size, __FILE__, __LINE__)

void cgreen_free(void* pointer, const char * filename, int line);
#define free(pointer) cgreen_free(pointer, __FILE__, __LINE__)

#else
#undef check_memory_leak
#define check_memory_leak(reporter)
#endif

#endif
