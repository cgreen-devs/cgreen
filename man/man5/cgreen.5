.mso www.tmac
.TH CGREEN 5
.SH NAME
cgreen \- a framework for C and C++ unit tests and mocking
.SH DESCRIPTION
.B Cgreen
is a framework for creating and running compact and easy-to-read
unittests for C and/or C++.

Here are some of its features:

.IP \[bu] 2
Fluent API resulting in very readable tests
.IP \[bu]
Expressive and clear output using the default reporter
.IP \[bu]
Fully functional mocks, both strict and loose
.IP \[bu]
Each test runs in its own process for test suite robustness
.IP \[bu]
Automatic discovery and running of tests using dynamic library inspection
.IP \[bu]
Extensive and expressive constraints for many datatypes
.IP \[bu]
BDD-flavoured test declarations with Before and After declarations
.IP \[bu]
Extensible reporting mechanism
.IP \[bu]
Fully composable test suites
.IP \[bu]
An isolated test can be run in a single process for debugging


.SH "SEE ALSO"
cgreen-runner(1)
.PP
The full documentation for
.B cgreen-runner
and
.B the Cgreen framework
is in the
.B Cgreen
manual available at
.URL https://github.com/cgreen-devs/cgreen GitHub .
