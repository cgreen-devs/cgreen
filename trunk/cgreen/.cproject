<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<?fileVersion 5.0.0?>

<cproject storage_type_id="org.eclipse.cdt.core.XmlProjectDescriptionStorage2">
	<storageModule moduleId="org.eclipse.cdt.core.settings">
		<cconfiguration id="cdt.managedbuild.toolchain.gnu.base.1521338569">
			<storageModule externalCElementFile="cdt.managedbuild.toolchain.gnu.base.1521338569_org.eclipse.cdt.core.settings" id="cdt.managedbuild.toolchain.gnu.base.1521338569" name="Default"/>
			<storageModule externalCElementFile="cdt.managedbuild.toolchain.gnu.base.1521338569_cdtBuildSystem" version="4.0.0"/>
			<storageModule externalCElementFile="cdt.managedbuild.toolchain.gnu.base.1521338569_org.eclipse.cdt.core.externalSettings"/>
		</cconfiguration>
	</storageModule>
	<storageModule moduleId="cdtBuildSystem" version="4.0.0">
		<project id="cgreen.null.1958735211" name="cgreen"/>
	</storageModule>
	<storageModule moduleId="org.eclipse.cdt.core.LanguageSettingsProviders"/>
	<storageModule moduleId="org.eclipse.cdt.make.core.buildtargets">
		<buildTargets>
			<target name="all" path="" targetID="org.eclipse.cdt.build.MakeTargetBuilder">
				<buildCommand>make</buildCommand>
				<buildArguments/>
				<buildTarget>all</buildTarget>
				<stopOnError>true</stopOnError>
				<useDefaultCommand>false</useDefaultCommand>
				<runAllBuilders>true</runAllBuilders>
			</target>
			<target name="clean" path="" targetID="org.eclipse.cdt.build.MakeTargetBuilder">
				<buildCommand>rm</buildCommand>
				<buildArguments>-rf build</buildArguments>
				<buildTarget/>
				<stopOnError>true</stopOnError>
				<useDefaultCommand>false</useDefaultCommand>
				<runAllBuilders>true</runAllBuilders>
			</target>
		</buildTargets>
	</storageModule>
	<storageModule moduleId="scannerConfiguration">
		<autodiscovery enabled="true" problemReportingEnabled="true" selectedProfileId=""/>
		<scannerConfigBuildInfo instanceId="cdt.managedbuild.toolchain.gnu.base.1521338569;cdt.managedbuild.toolchain.gnu.base.1521338569.1174655132;cdt.managedbuild.tool.gnu.c.compiler.base.349579178;cdt.managedbuild.tool.gnu.c.compiler.input.1942191604">
			<autodiscovery enabled="false" problemReportingEnabled="true" selectedProfileId="org.eclipse.cdt.make.core.GCCStandardMakePerFileProfile"/>
		</scannerConfigBuildInfo>
		<scannerConfigBuildInfo instanceId="cdt.managedbuild.toolchain.gnu.base.1521338569;cdt.managedbuild.toolchain.gnu.base.1521338569.1174655132;cdt.managedbuild.tool.gnu.cpp.compiler.base.1611546606;cdt.managedbuild.tool.gnu.cpp.compiler.input.560078280">
			<autodiscovery enabled="true" problemReportingEnabled="true" selectedProfileId=""/>
		</scannerConfigBuildInfo>
	</storageModule>
</cproject>
