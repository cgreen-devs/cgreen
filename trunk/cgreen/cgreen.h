#include "unit.h"
#include "text_reporter.h"
#include "assertions.h"
#include <stdlib.h>
