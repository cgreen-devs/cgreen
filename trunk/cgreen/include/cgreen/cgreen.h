#include <cgreen/unit.h>
#include <cgreen/text_reporter.h>
#include <cgreen/cute_reporter.h>
#include <cgreen/cdash_reporter.h>
#include <cgreen/assertions.h>
#include <stdlib.h>
