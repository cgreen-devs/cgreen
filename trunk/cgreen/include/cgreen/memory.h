#ifndef MEMORY_HEADER
#define MEMORY_HEADER
#ifdef __cplusplus
    extern "C" {
#endif

typedef struct MemoryPool_ MemoryPool;

MemoryPool *create_memory_pool();
void free_memory_pool(MemoryPool *pool);
void *memory_pool_allocate(MemoryPool *pool, size_t bytes);
void *memory_pool_reallocate(MemoryPool *pool, void *pointer, size_t bytes);

#ifdef __cplusplus
    }
#endif
#endif
