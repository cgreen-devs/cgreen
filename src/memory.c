#include <stdlib.h>
#include <string.h>
#include <include/cgreen/reporter.h>
#include <stdio.h>
#include "include/cgreen/memory.h"

typedef struct MemoryEntry_ {
    void * memory_location;
    size_t size;
    const char* filename;
    int line_number;
} MemoryEntry;


static MemoryEntry ** allocation_memory = NULL;
static int number_of_memory_entries = 0;


static int find_memory_position(void * pointer);
static void move_memory_back(int position_to_start_on);
static MemoryEntry * get_memory_entry(void * pointer);

#undef malloc
static void create_allocation_memory(void) {
    allocation_memory = malloc(sizeof(MemoryEntry)*100);
}
#define malloc cgreen_malloc

#undef malloc
void* cgreen_malloc(size_t size, const char * filename, int line) {
    if(allocation_memory == NULL)
        create_allocation_memory();

    void * new_memory_pointer = malloc(size);
    MemoryEntry * new_memory_entry = malloc(sizeof(MemoryEntry));
    new_memory_entry->memory_location = new_memory_pointer;
    new_memory_entry->size = size;
    new_memory_entry->filename = filename;
    new_memory_entry->line_number = line;
    allocation_memory[number_of_memory_entries] = new_memory_entry;
    number_of_memory_entries++;

    return new_memory_pointer;
}
#define malloc cgreen_malloc

void* cgreen_calloc(size_t num_members, size_t size, const char * filename, int line) {
    void * allocated_memory = cgreen_malloc(num_members * size, filename, line);
    if (allocated_memory)
        memset(allocated_memory, 0, num_members * size);
    return allocated_memory;
}

void* cgreen_realloc(void* pointer, size_t size, const char * filename, int line) {
    if (pointer == NULL)
        return cgreen_malloc(size, filename, line);

    if (size == 0) {
        cgreen_free(pointer, filename, line);
        return NULL;
    }

    MemoryEntry * memory_position = get_memory_entry(pointer);

    void * new_pointer = cgreen_malloc(size, filename, line);

    size_t actual_size = size;
    if (size > memory_position->size)
        actual_size = memory_position->size;

    memcpy(new_pointer, pointer, actual_size);

    cgreen_free(pointer, filename, line);

    return new_pointer;
}


#undef free
void cgreen_free(void * pointer, const char * filename, const int line) {
    va_list null_arguments;
    memset(&null_arguments, 0, sizeof(null_arguments));
    if (pointer == NULL) {

        TestReporter *reporter = get_test_reporter();
        reporter->show_fail(reporter, filename, line, "Trying to free memory using a NULL pointer", null_arguments);
        send_reporter_exception_notification(reporter);
        return;
    }

    int memory_position = find_memory_position(pointer);

    if (memory_position == -1) {
        char message[100];
        sprintf(message, "Trying to free memory [%p] that was not previously allocated", pointer);

        TestReporter *reporter = get_test_reporter();
        reporter->show_fail(reporter, filename, line, message, null_arguments);
        send_reporter_exception_notification(reporter);
        return;
    }

    free(allocation_memory[memory_position]);
    move_memory_back(memory_position);
    number_of_memory_entries--;
}
#define free cgreen_free

void check_memory_leak(TestReporter *reporter) {
    if (number_of_memory_entries > 0) {
        for( int i = 0 ; i < number_of_memory_entries ; i++) {
            MemoryEntry * memoryEntry = allocation_memory[i];
            (reporter->assert_true)(
                    reporter,
                    memoryEntry->filename,
                    memoryEntry->line_number,
                    0,
                    "Memory allocated [%p] was not freed [%s:%d]",
                    memoryEntry->memory_location,
                    memoryEntry->filename,
                    memoryEntry->line_number
            );
        }
    }
}

int find_memory_position(void * pointer) {
    for( int i = 0 ; i < number_of_memory_entries ; i++) {
        MemoryEntry * memoryEntry = allocation_memory[i];
        if(memoryEntry && memoryEntry->memory_location == pointer)
            return i;
    }

    return -1;
}

MemoryEntry * get_memory_entry(void * pointer) {
    int position = find_memory_position(pointer);

    if (position == -1)
        return NULL;

    return allocation_memory[position];
}

void move_memory_back(int position_to_start_on) {
    for(int i = position_to_start_on ; i < number_of_memory_entries - 1 ; i++) {
        allocation_memory[i] = allocation_memory[i+1];
    }
}