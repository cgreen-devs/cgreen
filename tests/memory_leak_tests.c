#include <cgreen/cgreen.h>
#include <cgreen/messaging.h>
#include <stdlib.h>
#include <stdio.h>
#include "include/cgreen/memory.h"

Describe(MemoryLeak)
BeforeEach(MemoryLeak) {}
AfterEach(MemoryLeak) {}

Ensure(MemoryLeak, free_null_pointer) {
    free(NULL);
}

Ensure(MemoryLeak, alloc_and_free_memory) {
    int* new_int = malloc(sizeof(int));
    free(new_int);
    check_memory_leak(get_test_reporter());
    assert_that(true, is_true);
}

Ensure(MemoryLeak, multiple_allocations_and_free_of_memory_in_different_order_works) {
    int* new_int = malloc(sizeof(int));
    int* new_int1 = malloc(sizeof(int));
    int* new_int2 = malloc(sizeof(int));

    free(new_int);

    int* new_int3 = malloc(sizeof(int));

    free(new_int2);
    free(new_int1);
    free(new_int3);

    check_memory_leak(get_test_reporter());
    assert_that(true, is_true);
}

Ensure(MemoryLeak, free_unallocate_memory) {
    free((void*)1);

    check_memory_leak(get_test_reporter());
}

Ensure(MemoryLeak, fail_test_when_memory_is_leaked) {
    malloc(sizeof(int));

    check_memory_leak(get_test_reporter());
}

Ensure(MemoryLeak, use_realloc_maintain_data_in_string) {
    char * new_pointer = malloc(sizeof(char) * 5);
    strcpy(new_pointer, "123");

    char* new_string = realloc(new_pointer, sizeof(char) * 10);
    assert_that(new_string, is_equal_to_string("123"));

    free(new_string);
}

Ensure(MemoryLeak, use_realloc_to_increase_the_size_of_a_pointer) {
    char * new_pointer = malloc(sizeof(char) * 5);

    char* new_string = realloc(new_pointer, sizeof(char) * 10);
    strcpy(new_string, "123456789");
    free(new_string);
    check_memory_leak(get_test_reporter());
}

Ensure(MemoryLeak, use_realloc_to_decrease_the_size_of_a_pointer) {
    char * new_pointer = malloc(sizeof(char) * 10);

    char* new_string = realloc(new_pointer, sizeof(char) * 5);
    strcpy(new_string, "1234");

    free(new_string);
    check_memory_leak(get_test_reporter());
}

Ensure(MemoryLeak, use_calloc_to_allocate_an_array) {
    int * array = calloc(5, sizeof(int));

    array[0] = 1;
    array[1] = 1;
    array[2] = 1;
    array[3] = 1;
    array[4] = 1;

    free(array);
    check_memory_leak(get_test_reporter());
}
