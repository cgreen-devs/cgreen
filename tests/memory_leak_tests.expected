Running "memory_leak_tests" (9 tests)...
memory_leak_tests.c: Failure: MemoryLeak -> fail_test_when_memory_is_leaked 
	Memory allocated [...] was not freed [memory_leak_tests.c:46]

memory_leak_tests.c: Failure: MemoryLeak -> free_null_pointer 
	Trying to free memory using a NULL pointer

memory_leak_tests.c: Failure: MemoryLeak -> free_unallocate_memory 
	Trying to free memory [0x1] that was not previously allocated

  "MemoryLeak": 3 passes, 1 failure, 2 exceptions in 0ms.
Completed "memory_leak_tests": 3 passes, 1 failure, 2 exceptions in 0ms.
